lib/wampproto/session.rb

```diff
@@ -0,0 +1,113 @@
+# frozen_string_literal: true
+
+module Wampproto
+  # Session
+  class Session
+    attr_reader :serializer
+
+    def initialize(serializer = Serializer::JSON)
+      @serializer = serializer
+      init_state
+    end
+
+    ACCESSORS = %i[
+      call_requests
+      register_requests
+      registrations
+      invocation_requests
+      unregister_requests
+
+      publish_requests
+      subscribe_requests
+      subscriptions
+      unsubscribe_requests
+    ].freeze
+
+    def send_message(msg) # rubocop:disable Metrics/MethodLength, Metrics/AbcSize, Metrics/CyclomaticComplexity
+      case msg
+      when Message::Call
+        call_requests[msg.request_id] = msg.request_id
+      when Message::Register
+        register_requests[msg.request_id] = msg.request_id
+      when Message::Unregister
+        unregister_requests[msg.request_id] = msg.registration_id
+      when Message::Yield
+        unless invocation_requests.include?(msg.request_id)
+          raise StandardError, "cannot yield for unknown invocation request"
+        end
+
+        invocation_requests.delete(msg.request_id)
+      when Message::Publish
+        publish_requests[msg.request_id] = msg.request_id if msg.options.fetch(:acknowledge, false)
+      when Message::Subscribe
+        subscribe_requests[msg.request_id] = msg.request_id
+      when Message::Unsubscribe
+        unsubscribe_requests[msg.request_id] = msg.subscription_id
+      end
```

Comment:
any other message here should raise an exception, because there is only a handful of messages that a client can send and the above covers them all